src/control/cmd/daos_admin/handler.go

```diff
@@ -0,0 +1,149 @@
+//
+// (C) Copyright 2019 Intel Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+// The Government's rights to use, modify, reproduce, release, perform, display,
+// or disclose this software are subject to the terms of the Apache License as
+// provided in Contract No. 8F-30005.
+// Any reproduction of computer software, computer software documentation, or
+// portions thereof marked with this legend must also reproduce the markings.
+//
+package main
+
+import (
+	"encoding/json"
+	"io"
+
+	"github.com/pkg/errors"
+
+	"github.com/daos-stack/daos/src/control/logging"
+	"github.com/daos-stack/daos/src/control/pbin"
+	"github.com/daos-stack/daos/src/control/server/storage/scm"
+)
+
+func sendFailure(err error, res *pbin.Response, dest io.Writer) error {
+	res.Error = &pbin.RequestFailure{Message: err.Error()}
+	data, err := json.Marshal(res)
+	if err != nil {
+		return err
+	}
+
+	_, err = dest.Write(data)
+	return err
+}
+
+func sendSuccess(payloadSrc interface{}, res *pbin.Response, dest io.Writer) error {
+	payload, err := json.Marshal(payloadSrc)
+	if err != nil {
+		return sendFailure(err, res, dest)
+	}
+
+	res.Payload = payload
+	data, err := json.Marshal(res)
+	if err != nil {
+		return err
+	}
+
+	_, err = dest.Write(data)
+	return err
+}
+
+func readRequest(log logging.Logger, rdr io.Reader) (*pbin.Request, error) {
+	buf := make([]byte, pbin.MaxMessageSize)
+
+	readLen, err := rdr.Read(buf)
+	if err != nil {
+		return nil, err
+	}
+
+	var req pbin.Request
+	if err := json.Unmarshal(buf[:readLen], &req); err != nil {
+		return nil, err
+	}
+
+	return &req, nil
+}
+
+func handleRequest(log logging.Logger, scmProvider *scm.Provider, req *pbin.Request, resDest io.Writer) (err error) {
```

Comment:
Would prefer not to, actually. The current implementation allows for agile development and minimal maintenance burden by using JSON as the wire representation over stdio rather than imposing the overhead of protobuf definitions and unix sockets. We also discussed the idea of dRPC for the privileged binary in the f2f and in subsequent design discussions and decided that we didn't want the privileged binary to be able to receive input from anything other than its parent process.